docs/conf.py

```diff
@@ -155,3 +160,47 @@
 # ----------------------------------------------------------------------------------
 
 plot_html_show_formats = False
+
+
+# ----------------------------------------------------------------------------------
+# Source code links
+# ----------------------------------------------------------------------------------
+
+def linkcode_resolve(domain, info):
+    if domain != "py":
+        return None
+
+    module_name = info["module"]
+    if "qiskit" not in module_name:
+        return None
+     
+    try: 
+        module = importlib.import_module(module_name)
+    except ModuleNotFoundError:
+        return None
+
+    obj = module
+    for part in info["fullname"].split("."):
+        try:
+            obj = getattr(obj, part)
+        except AttributeError:
+            return None
+
+    try:
+        full_file_name = inspect.getsourcefile(obj)
+    except TypeError:
+        return None
+    if full_file_name is None:
+        return None
+    file_name = full_file_name.split("/qiskit/")[-1]
```

Comment:
It's quite unlikely to matter, but you might prefer to do this `full_file_name.split` thing using the proper `os.path` or `pathlib` tooling.  Something like:

```python
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]

def linkcode_resolve(domain, info):
    # ...

    file_name = str(Path(full_file_name).resolve().relative_to(REPO_ROOT / "qiskit"))
```

The current form would break if we were ever to add a subdirectory of the `qiskit` package that itself was called `qiskit` - unlikely, but not impossible.

Reply:
Oh, actually not even unlikely - I just remembered that I literally proposed doing that in #10737 (though in that case, there wouldn't be any Python-documentable objects in it).

Reply:
Thanks for the suggestion!

Reply:
Ah, I now remember why we were using `[-1]`. This current code results in the relative file path `.tox/docs/lib/python3.8/site-packages/qiskit/circuit/classicalfunction/exceptions.py`. 

I'm unclear why that is: I couldn't reproduce in `qiskit_sphinx_theme` with setting in `tox.ini` `isolated_build = true` and `usedevelop = false` to mirror Qiskit. (I've been having issues running tox locally in Qiskit due to autodoc). In `qiskit_sphinx_theme` locally and GitHub Actions, the path is the normal file path you'd expect without the `.tox` prefix. So I think we want this code to support both styles.

We could use a regex replace to remove the `.tox` prefix if it's there.

Reply:
In that case, perhaps you want to use `import qiskit; qiskit.__file__` as the file-path root, so it'll automatically find the base of where the files are installed?

I don't mind a huge amount, though.